### PR TITLE
reduce turn length to 1 second

### LINKED ISSUE #30018: Reduce turn length to 1 second
**Problem**
The status quo of 6 second turns effectively gives us days shorter than they should be and makes simple actions take a disproportionately long time compared to long actions. Some situations like spending an hour walking across a small town and 6 second melee attacks are quite jarring.

**Solution**  
The length of a turn should be reduced to 1 second. This would reduce a tick to 10ms, providing good granularity for very short actions.

A lot of the implementation should fall into place if tick/turn conversions are encapsulated. Any json adjustments required across a large number of objects could be automated.

**Additional context**  
This would go well with the 91 day seasons change to set characters actions in a realistic timeframe.

### PR DESCRIPTION
**Pull request by ymber**
#### Summary
```SUMMARY: Balance "Reduce turn time to 1 second"```

#### Purpose of change
closes #30018 

#### Describe the solution
Tick to time conversions were already encapsulated, constants just needed changing.

- [x] change constants in conversions
- [x] update json values and documentation
- [x] fix construction times
- [x] fix times in player_activity
- [x] check behavior of max_duration

### CODE REVIEW

**Comment by Vasyan2006** [comment]
Crossbow reloading time is calculated in moves instead of second and must be updated for new turn duration. Same for many other actions.

**Comment by mlangsdorf** [comment]
https://github.com/CleverRaven/Cataclysm-DDA/blob/f3a679e222770513e96188b772b0dc1d8bb4bfd0/src/vehicle.cpp#L4140 needs to be changed to
```diff
-        consume_fuel( idle_rate, 6 * to_turns<int>( 1_turns ), true );
+        consume_fuel( idle_rate, to_turns<int>( 1_turns ), true );
```
There are other places in the vehicle code that do a similar thing, and there are other locations in the code that need to be audited for similar uses of to_turns.

**Comment by mlangsdorf** [comment]
You'll need this to adjust the vehicle efficiency test for the change in idle consumption.
```diff
diff --git a/tests/vehicle_efficiency.cpp b/tests/vehicle_efficiency.cpp
index f71e151707..f38b9e9664 100644
--- a/tests/vehicle_efficiency.cpp
+++ b/tests/vehicle_efficiency.cpp
@@ -430,22 +430,22 @@ TEST_CASE( "vehicle_make_efficiency_case", "[.]" )
 // Fix test for electric vehicles
 TEST_CASE( "vehicle_efficiency", "[vehicle] [engine]" )
 {
-    test_vehicle( "beetle", 767373, 235000, 195500, 70760, 54070 );
-    test_vehicle( "car", 1072322, 363000, 245700, 44380, 26120 );
-    test_vehicle( "car_sports", 1098408, 208900, 168600, 35020, 20180 );
+    test_vehicle( "beetle", 767373, 294400, 235600, 77390, 58340 );
+    test_vehicle( "car", 1072322, 530700, 312700, 47890, 28180 );
+    test_vehicle( "car_sports", 1098408, 456500, 297800, 39400, 21740 );
     test_vehicle( "electric_car", 1070791, 212500, 121700, 17430, 9118 );
-    test_vehicle( "suv", 1271990, 696200, 405800, 68340, 32700 );
-    test_vehicle( "motorcycle", 162785, 64180, 54160, 41250, 34600 );
-    test_vehicle( "quad_bike", 264745, 63520, 63520, 31390, 31390 );
-    test_vehicle( "scooter", 62287, 140400, 138000, 108800, 107000 );
-    test_vehicle( "superbike", 241785, 58540, 40000, 30480, 19890 );
-    test_vehicle( "ambulance", 1783889, 331200, 281700, 61580, 46220 );
-    test_vehicle( "fire_engine", 2563241, 1018000, 863500, 236700, 208700 );
-    test_vehicle( "fire_truck", 6259233, 253000, 149700, 19010, 4523 );
-    test_vehicle( "truck_swat", 5939334, 409300, 248500, 27990, 7189 );
-    test_vehicle( "tractor_plow", 703658, 354200, 354200, 106700, 106700 );
-    test_vehicle( "apc", 5753619, 957100, 824200, 123600, 85540 );
-    test_vehicle( "humvee", 5475145, 465800, 278700, 24650, 8824 );
-    test_vehicle( "road_roller", 8779702, 315600, 295100, 21910, 6666 );
+    test_vehicle( "suv", 1271990, 998800, 492900, 71890, 34300 );
+    test_vehicle( "motorcycle", 162785, 76870, 62950, 47530, 38890 );
+    test_vehicle( "quad_bike", 264745, 75950, 75950, 35080, 35080 );
+    test_vehicle( "scooter", 62287, 266900, 258300, 195200, 189700 );
+    test_vehicle( "superbike", 241785, 72120, 46710, 34580, 21720 );
+    test_vehicle( "ambulance", 1783889, 433100, 351300, 65530, 48600 );
+    test_vehicle( "fire_engine", 2563241, 1171000, 970800, 248500, 219000 );
+    test_vehicle( "fire_truck", 6259233, 308500, 200800, 19950, 4747 );
+    test_vehicle( "truck_swat", 5939334, 505800, 349500, 30060, 7719 );
+    test_vehicle( "tractor_plow", 703658, 528000, 528000, 121900, 121900 );
+    test_vehicle( "apc", 5753619, 1100000, 942800, 132800, 91850 );
+    test_vehicle( "humvee", 5475145, 607600, 342700, 25880, 9263 );
+    test_vehicle( "road_roller", 8779702, 369800, 412000, 22990, 6996 );
     test_vehicle( "golf_cart", 446230, 52800, 104200, 26830, 13890 );
 };
```

**Comment by Hirmuolio** [comment]
`X_turns` is used for duration of various things in many places. Almost all of them need to be changed. 

Probably best to change many of them them to `X_seconds` instead of 6x`X_turns` unless they specifically are about turns instead of just time duration.

**Comment by mlangsdorf** [comment]
I played the current version for a bit - long enough to loot about 7 houses, drive about 100 overmap tiles back to my base, unload my loot, and do some basic maintenance on my ride.  It wasn't a comprehensive test, but a decent start.

I didn't notice any glaring errors or problems.  The looting portion of the game plays *much* faster, obviously, which may make the "instant expert" problem more severe.

I need to review the vehicle braking code.  Hand-waving the rate of deceleration and skidding was probably acceptable when the turn length was 6 seconds, but now that the turn length is 1 second, it shouldn't be possible to brake from 60 mph in a single second.  I don't think that should hold up this PR; it's just something to consider in the future.

**Comment by ymber** [comment]
As far as i can tell whoever wrote item::calc_temp in the first place multiplied the conductivity constant of air by 6 so he could use 6 second turns as a time unit then used that constant as the heat transfer coefficient in the formula for conductivity_term. It sounds plausible and changing the constant accordingly made the tests pass so that's what I'm going with. conductivity_constant now depends on actual watts and not six second watts.

File: src/addiction.cpp [hunk 1]
```diff
@@ -116,7 +116,7 @@ void addict_effect( player &u, addiction &add )
             if( one_in( 20 ) && dice( 2, 20 ) < in ) {
                 u.add_msg_if_player( m_bad, _( "Your hands start shaking... you need some painkillers." ) );
                 u.add_morale( MORALE_CRAVING_OPIATE, -40, -10 * in );
-                u.add_effect( effect_shakes, 2_minutes + in * 5_turns );
+                u.add_effect( effect_shakes, 2_minutes + in * 30_turns );
```
Comment (Night-Pryanik) [inline src/addiction.cpp]: 30_seconds or 5_seconds?

**Comment by Firestorm01X2** [comment]
It is recommended to check aiming speed in ranged.cpp.

**Comment by ymber** [comment]
Aim speed should be ok, it's all measured in moves.

File: src/addiction.cpp (hunk 1 above)
Comment (ymber) [reply src/addiction.cpp]: 30_seconds. Missed that when I was changing them.

**Comment by mlangsdorf** [comment]
I just realized that you're missing construction recipes.  I thought I built that fireplace unreasonably quickly!

**Comment by kevingranade** [comment]
Since most of this PR is changing crafting recipe time fields, what do you think about having "old style" time entries remain 6 seconds, and changing everything in the main repository to be like ``"time": "5 sec"`` or similar?

**Comment by mlangsdorf** [comment]
player_activity is also broken - I'm chopping down trees in 15 minutes.

It looks like you need to go through iuse.cpp, iuse_actor.cpp, and iexamine.cpp and audit all moves - some things are meant to take X turns and can stay how they are, but other things should take Y seconds.  Everything should use `to_moves()`

**Comment by ymber** [comment]
Changing the units of time fields is an option. I was thinking of making it like how volume fields are handled with different units.

**Comment by Rivet-the-Zombie** [comment]
It would be pretty nice to not have to spend twenty seconds aiming at a target in order to actually stand a chance of hitting it.

File: src/iuse.cpp [hunk 2]
```diff
@@ -7966,7 +7966,7 @@ washing_requirements washing_requirements_for_volume( units::volume vol )
 {
     int water = divide_round_up( vol, 125_ml );
     int cleanser = divide_round_up( vol, 1000_ml );
-    int time = 1000 * vol / 250_ml;
+    int time = to_turns<int>( 10_seconds * ( vol / 250_ml ) );
```
Comment (Night-Pryanik) [inline src/iuse.cpp]:
```suggestion
    int time = to_turns<int>( 10_seconds * vol / 250_ml );
```
These brackets are redundant.

Comment (ymber) [reply src/iuse.cpp]: Clang builds break without the brackets.
```
clang++  -DRELEASE -DGIT_VERSION -DTILES -DBACKTRACE -ffast-math -Os -Wall -Wextra -fsigned-char -Werror -std=c++11 -MMD -MP -I/usr/include/SDL2 -D_REENTRANT -I/usr/include/SDL2 -D_REENTRANT -I/usr/include/SDL2 -D_REENTRANT -c src/iuse_actor.cpp -o obj/tiles/iuse_actor.o
In file included from src/iuse.cpp:20:
In file included from src/artifact.h:8:
In file included from src/itype.h:12:
src/calendar.h:440:20: error: no matching conversion for functional-style cast from 'quantity<decltype(std::declval<int>() * std::declval<int>()), units::volume_in_milliliter_tag>'
      (aka 'quantity<int, units::volume_in_milliliter_tag>') to 'time_duration'
            return time_duration( lhs.turns_ * rhs );
                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/iuse.cpp:7969:42: note: in instantiation of function template specialization 'operator*<units::quantity<int, units::volume_in_milliliter_tag> >' requested here
    int time = to_turns<int>( 10_seconds * vol / 250_ml );
                                         ^
src/calendar.h:300:7: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'quantity<decltype(std::declval<int>() * std::declval<int>()),
      units::volume_in_milliliter_tag>' (aka 'quantity<int, units::volume_in_milliliter_tag>') to 'const time_duration' for 1st argument
class time_duration
      ^
src/calendar.h:300:7: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'quantity<decltype(std::declval<int>() * std::declval<int>()),
      units::volume_in_milliliter_tag>' (aka 'quantity<int, units::volume_in_milliliter_tag>') to 'time_duration' for 1st argument
class time_duration
      ^
src/calendar.h:306:28: note: candidate constructor not viable: no known conversion from 'quantity<decltype(std::declval<int>() * std::declval<int>()), units::volume_in_milliliter_tag>'
      (aka 'quantity<int, units::volume_in_milliliter_tag>') to 'const int' for 1st argument
        explicit constexpr time_duration( const int t ) : turns_( t ) { }
                           ^
src/calendar.h:312:9: note: candidate constructor not viable: no known conversion from 'quantity<decltype(std::declval<int>() * std::declval<int>()), units::volume_in_milliliter_tag>'
      (aka 'quantity<int, units::volume_in_milliliter_tag>') to 'const std::nullptr_t' (aka 'const nullptr_t') for 1st argument
        time_duration( const std::nullptr_t ) : turns_( 0 ) { }
        ^
src/calendar.h:309:9: note: candidate constructor not viable: requires 0 arguments, but 1 was provided
        time_duration() = default;
        ^
src/calendar.h:432:46: error: overload resolution selected deleted operator '/'
            return time_duration( lhs.turns_ / rhs );
                                  ~~~~~~~~~~ ^ ~~~
src/iuse.cpp:7969:48: note: in instantiation of function template specialization 'operator/<units::quantity<int, units::volume_in_milliliter_tag> >' requested here
    int time = to_turns<int>( 10_seconds * vol / 250_ml );
                                               ^
src/units.h:215:13: note: candidate function [with lvt = int, ut = units::volume_in_milliliter_tag, rvt = int, $3 = void] has been explicitly deleted
inline void operator/( lvt, quantity<rvt, ut> ) = delete;
            ^
src/calendar.h:268:25: note: candidate function template not viable: no known conversion from 'const int' to 'const time_duration' for 1st argument
constexpr time_duration operator/( const time_duration &lhs, const T rhs );
                        ^
src/units.h:208:1: note: candidate template ignored: could not match 'quantity<type-parameter-0-0, type-parameter-0-1>' against 'const int'
operator/( const quantity<lvt, ut> &lhs, const rvt &divisor )
^
src/units.h:220:1: note: candidate template ignored: could not match 'quantity<type-parameter-0-0, type-parameter-0-1>' against 'const int'
operator/( const quantity<lvt, ut> &lhs, const quantity<rvt, ut> &rhs )
^
2 errors generated.
make: *** [Makefile:768: obj/tiles/iuse.o] Error 1
make: *** Waiting for unfinished jobs....
```

Comment (Night-Pryanik) [reply src/iuse.cpp]: Ok, got it.

**Comment by Mcrone617** [comment]
Hey @ymber i'm working with @DracoGriffin to try to increase public participation in the development process, specifically by posting PR testing on reddit, and having people test out big changes that need a lot of testing to get merged. Would you be interested in having your PR as a public test? All you have to do is make a downloadable build and release it on GitHub, ( I can help you with that if you don't know how) and then i'll take it from there, and make a post describing your changes, and what people should do to test them and leave feedback. Of course you're welcome to post it yourself. Because this PR is such a big balance thing, I think it could really benefit from this, even 3-4 people saying that X is fine, or Y needs some fine tuning can really help in the merging process. Let me know if you're interested, i'm active here, and on the discord. I hope to hear back from you!

**Comment by kevingranade** [comment]
We can pull all the arbitrary time units to real time units changes in a seperate PR so those can land without changing behavior while the last parts are hunted down.

**Comment by ymber** [comment]
@Mcrone617 if you want to do testing builds from my feature branch go for it but I don't want to handle builds and releases because I work on linux and I don't want to have to handle cross compiling for windows players. Also this PR is getting split up so I can convert and standardize units for all the time fields before handling everything else, so you might not get much mileage out of testing this in its current state.

**Comment by ymber** [comment]
Rebased all this on top of the json time string changes. Should be ready now unless something else comes up.

**Comment by Hirmuolio** [comment]
you'll need to do something to the zombie reviving.
in its current state it is sneakily dependent on the turn length. 
Every turn it rolls a random number to check if the zombie revives. Doing this more frequently can drastically speed up zombie reviving.

**Comment by Mooses2k** [comment]
If looting is going too quickly, maybe add time when going through dressers and wardrobes and such. Make them like safes that are unlocked but take time to search before you can see the contents? Adds realism since I shouldn't be able to stand next to a dresser and magically divine it's contents instantly.

**Comment by ymber** [comment]
item_counter is and the item::process* functions look ok. Zombies reviving looks like it's intended to measure in turns irrespective of what that timeframe really is. item::process_tool is being updated - see #30656.

**Comment by Hirmuolio** [comment]
About the zombie reviving.

Lets assume we have a zombie corpse that has 1% chance to revive.  
You stand next to the corpse for one minute.

With 6 s turns there is 10% chance that the corpse revives during that minute ( 1-0.99^10 ).
With 1 s turns there is 45% chance that the corpse revives during that minute ( 1-0,99^60 ).

**Comment by kevingranade** [comment]
This is so incredibly close.
@Hirmuolio is right about respawns, but I'm not terribly concerned about that, because it still doesn't rise for 6 hours, and then the rate of respawn climbs as expected as time since death approaches 48 hrs.
There's an overflow error in an item aging test, item ages wrap after a year now.
This definitely missed something in nutrition handling, vitamin handling is off by a factor of 6.

**Comment by Hirmuolio** [comment]
Just dividing the probablitiy by 6 would probably be enough. It won't be exactly same as before but should be close enough.

**Comment by kevingranade** [comment]
Prospective fixes:
#### Respawn rate
drop a
```
if( !calendar::once_every( 6_seconds ) ) {
    return false;
}```
around here https://github.com/CleverRaven/Cataclysm-DDA/blob/02eaddfe851a5b142644fc3d2cbdd4bb09c91d04/src/item.cpp#L4331

#### Overflow Error
I *thought* this was a symptom of the underlying data in time_point overflowing, but it claims it is an int (so 31 bits of precision) representing number of turns, so I don't know why it would have overflowed, maybe there's something weird happening in the test...

#### Vitamin discrepancy
The nutrition tick is normalized for time since last tick, so it is working as expected, need to drill down into the vitamin handling, it might be doing something strange.

**Comment by kevingranade** [comment]
I've been trying to trace where the apparent overflow was coming from, and discovered it's actually an underflow.  The item used was necco wafers, which spoil after 180 days, and then we modify their age by spoilage time - one year, which as it turns out resolves to a very negative number, so of course it doesn't work.  I'm not sure now why this ever passed, but now I suspect this is just a red herring, if we swap in an item with a spoilage time longer than a year it should just work.

**Comment by kevingranade** [comment]
Aha, the vitamin problem is pretty straightforward. Vitamin change rates are specified in turns:
https://github.com/CleverRaven/Cataclysm-DDA/blob/master/data/json/vitamin.json
Since the list is short and stable, I'd support either editing the json values directly or extending the parser to handle durations with units like elsewhere.

**Comment by Fukken-Saved** [comment]
The primary effect of reducing turn duration seems to be that now the taking forever to craft items problem is six times as bad, so as of now it's unplayable unless you can move it to machine language or something so it's six times as fast.
I like having all z-levels simulated so if I wanted to do that there's probably some version I can use before this update (not the awful one where the stamina was all messed up), it would be much better if turn duration could be player adjusted
Also, realistically as far as zombie sims go nobody is going to be able to run through a house and in 30 seconds grab a hundred pounds and 50 liters of very specific manually selected items that they want, the net effect of this is that it's going to make day-to-day survival wayyy easier as you effectively have days 6 times as long

**Comment by Fukken-Saved** [comment]
Also, everyone is going to be bored of it probably before even the first Summer

**Comment by kevingranade** [comment]
> realistically as far as zombie sims go nobody is going to be able to run through a house and in 30 seconds grab a hundred pounds and 50 liters of very specific manually selected items that they want,

This is a first coarse adjustment, all action durations can be tweaked to make them more reasonable, but it's now much closer than it was before.

**Comment by ZhilkinSerg** [comment]
> 
> 
> The primary effect of reducing turn duration seems to be that now the taking forever to craft items problem is six times as bad, so as of now it's unplayable

That is not true.

**Comment by Fukken-Saved** [comment]
I forgot where this was, but I just want to point out that build 9112 seems to be the last version before the new turn durations take effect. Chances are you'll be able to adjust turn duration eventually, but I've had very good experience with the build with most of the crashes on worldgen
And it is unplayable if you have to wait 10 minutes while it's crafting a stone knife.

**Pull request merged**